src/QsCompiler/CompilerOptimizations/ExpressionEvaluation.fs

```diff
@@ -63,32 +65,43 @@ and [<AbstractClass>] internal ExpressionKindEvaluator(vars: VariablesDict, cd:
 
     override this.onIdentifier (sym, tArgs) =
         match sym with
-        | LocalVariable name -> vars.getVar name.Value |? Identifier (sym, tArgs)
+        | LocalVariable name -> getVar !stateRef name.Value |? Identifier (sym, tArgs)
         | _ -> Identifier (sym, tArgs)
 
     override this.onFunctionCall (method, arg) =
         let method, arg = this.simplify (method, arg)
-        if maxRecursiveDepth > 0 && isLiteral arg.Expression cd then
+        maybe {
             match method.Expression with
             | Identifier (GlobalCallable qualName, types) ->
-                fe.evaluateFunction qualName arg types |? CallLikeExpression (method, arg)
+                do! check (maxRecursiveDepth > 0 && isLiteral !stateRef arg.Expression)
+                return! fe.evaluateFunction !stateRef qualName arg types
             | CallLikeExpression (baseMethod, partialArg) ->
-                this.Transform (partialApplyFunction baseMethod partialArg arg)
-            | _ ->
-                failwithf "Unknown function call: %O" (printExpr method.Expression)
-        else CallLikeExpression (method, arg)
+                do! hasMissingExprs partialArg |> check
```

Comment:
The line "do! check (hasMissingExprs partialArg)" should have the same effect as checking TypedExpression.IsPartialApplication. When should we prefer one over the other?

Reply:
I know it has the same effect, but my suggestion is to pipe it through the proper override to make sure code only ever needs to be adapted in one place instead of two.

Reply:
Fixed in future PR